test/realm_set_test.dart

```diff
@@ -0,0 +1,535 @@
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////////
+
+import 'dart:typed_data';
+
+import 'package:test/test.dart' hide test, throws;
+import '../lib/realm.dart';
+
+import 'test.dart';
+
+part 'realm_set_test.g.dart';
+
+class _NullableBool {}
+
+class _NullableInt {}
+
+class _NullableString {}
+
+class _NullableDouble {}
+
+class _NullableDateTime {}
+
+class _NullableObjectId {}
+
+class _NullableUuid {}
+
+class _NullableObjects {}
+
+/// When changing update also `setByType`
+List<Type> supportedTypes = [
+  bool,
+  int,
+  String,
+  double,
+  DateTime,
+  ObjectId,
+  Uuid,
+  RealmValue,
+  RealmObject,
+  _NullableBool,
+  _NullableInt,
+  _NullableString,
+  _NullableDouble,
+  _NullableDateTime,
+  _NullableObjectId,
+  _NullableUuid
+];
+
+@RealmModel()
+class _Car {
+  @PrimaryKey()
+  late String make;
+}
+
+@RealmModel()
+class _TestRealmSets {
+  @PrimaryKey()
+  late int key;
+
+  late Set<bool> boolSet;
+  late Set<int> intSet;
+  late Set<String> stringSet;
+  late Set<double> doubleSet;
+  late Set<DateTime> dateTimeSet;
+  late Set<ObjectId> objectIdSet;
+  late Set<Uuid> uuidSet;
+  late Set<RealmValue> mixedSet;
+  late Set<_Car> objectsSet;
+
+  late Set<bool?> nullableBoolSet;
+  late Set<int?> nullableIntSet;
+  late Set<String?> nullableStringSet;
+  late Set<double?> nullableDoubleSet;
+  late Set<DateTime?> nullableDateTimeSet;
+  late Set<ObjectId?> nullableObjectIdSet;
+  late Set<Uuid?> nullableUuidSet;
+
+  /// When changing update also `supportedTypes`
+  Sets setByType(Type type) {
+    switch (type) {
+      case bool:
+        return Sets(boolSet as RealmSet<bool>, [true, false]);
+      case int:
+        return Sets(intSet as RealmSet<int>, [-1, 0, 1]);
+      case String:
+        return Sets(stringSet as RealmSet<String>, ['Tesla', 'VW', 'Audi']);
+      case double:
+        return Sets(doubleSet as RealmSet<double>, [-1.1, 0.1, 1.1, 2.2, 3.3, 3.14]);
+      case DateTime:
+        return Sets(dateTimeSet as RealmSet<DateTime>, [DateTime(2023).toUtc(), DateTime(1981).toUtc()]);
+      case ObjectId:
+        return Sets(objectIdSet as RealmSet<ObjectId>, [ObjectId.fromTimestamp(DateTime(2023).toUtc()), ObjectId.fromTimestamp(DateTime(1981).toUtc())]);
+      case Uuid:
+        return Sets(uuidSet as RealmSet<Uuid>, [Uuid.fromString("12345678123456781234567812345678"), Uuid.fromString("82345678123456781234567812345678")]);
+      case RealmValue:
+        return Sets(mixedSet as RealmSet<RealmValue>, [RealmValue.nullValue(), RealmValue.bool(true), RealmValue.int(1), RealmValue.string("text")]);
```

Comment:
It'll be a good idea to also test with a RealmValue containing a RealmObject.

Reply:
good catch. This unraveled a lot of issues. Fixed in the latest commit